group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -1083,6 +1094,11 @@ public void run() {
         public void complete(Throwable exception) {
             if (exception == null) {
                 future.complete(null);
+
+                if (writeTimeoutTask != null) {
```

Comment:
I think that we should cancel it whenever the event is completed, regardless of whether it is successful or not. In both cases, it is completed so the timer has no reason to continue. What do you think?

Reply:
i originally had it only when it completes successfully because TimerTask#cancel requires synchronization on SystemTimer. I guess we will need to get a lock on it anyways when it expires anyways.